### PR TITLE
Feat/relative object position

### PR DESCRIPTION
**Pull request by kosuke55**
## Types of PR

- [x] New Features
- [ ] Upgrade of existing features
- [ ] Bugfix

## Description

Support RelativeObjectPosition

NPC1: spawn with LanePosition
NPC2: spawn with RelativeObjectPosition (bigger than NPC1)
Success condition: the euclidean distance between NPC1 and NPC2 is almost same(less than 0.01)

https://github.com/tier4/scenario_simulator_v2/assets/39142679/c4e0cfdf-0de6-4137-9ea7-233b441d20bf

### CODE REVIEW

File: openscenario/openscenario_interpreter/include/openscenario_interpreter/syntax/relative_object_position.hpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+#include <openscenario_interpreter/syntax/entity_ref.hpp>
+#include <openscenario_interpreter/syntax/orientation.hpp>
+#include <pugixml.hpp>
+#include <traffic_simulator_msgs/msg/lanelet_pose.hpp>
+
+namespace openscenario_interpreter
+{
+inline namespace syntax
+{
+/* ---- RelativeObjectPosition --------------------------------------------------
```
Comment (HansRobo) [inline openscenario/openscenario_interpreter/include/openscenario_interpreter/syntax/relative_object_position.hpp:30]: Please put OpenSCENARIO version after element name like below
https://github.com/tier4/scenario_simulator_v2/blob/master/openscenario/openscenario_interpreter/include/openscenario_interpreter/syntax/deterministic_multi_parameter_distribution.hpp#L26

Comment (kosuke55) [reply openscenario/openscenario_interpreter/include/openscenario_interpreter/syntax/relative_object_position.hpp]: fixed in 
[fb34fc1](https://github.com/tier4/scenario_simulator_v2/pull/1030/commits/fb34fc1db23e879f1b8d3426acadf8c31b9f17cb)

File: openscenario/openscenario_interpreter/include/openscenario_interpreter/syntax/relative_object_position.hpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ *    <xsd:attribute name="dy" type="Double" use="required"/>
+ *    <xsd:attribute name="dz" type="Double" use="optional"/>
+ *  </xsd:complexType>
+ *
+ * -------------------------------------------------------------------------- */
+struct RelativeObjectPosition : private SimulatorCore::CoordinateSystemConversion
+{
+  const Orientation orientation;
+
+  const EntityRef reference;
```
Comment (yamacir-kit) [inline openscenario/openscenario_interpreter/include/openscenario_interpreter/syntax/relative_object_position.hpp:47]:
```suggestion
  const String entity_ref;
```

File: openscenario/openscenario_interpreter/src/syntax/position.cpp [hunk 3]
```diff
@@ -26,7 +26,7 @@ Position::Position(const pugi::xml_node & node, Scope & scope)
     choice(node,
       std::make_pair(         "WorldPosition", [&](auto && node) { return make<        WorldPosition>(node, scope); }),
       std::make_pair( "RelativeWorldPosition", [&](auto && node) { return make<RelativeWorldPosition>(node, scope); }),
-      std::make_pair("RelativeObjectPosition", [&](auto && node) { throw UNSUPPORTED_ELEMENT_SPECIFIED(node.name()); return unspecified; }),
+      std::make_pair("RelativeObjectPosition", [&](auto && node) { return make<RelativeObjectPosition>(node, scope); }),
```
Comment (yamacir-kit) [inline openscenario/openscenario_interpreter/src/syntax/position.cpp:29]: Adjust the column alignment according to the style of the original code. Do not try to break the style to reduce diffs.

File: openscenario/openscenario_interpreter/src/syntax/position.cpp [hunk 4]
```diff
@@ -43,6 +43,7 @@ Position::operator geometry_msgs::msg::Pose() const
       // clang-format off
       [&](const         WorldPosition & position) { return static_cast<geometry_msgs::msg::Pose>(position); },
       [&](const RelativeWorldPosition & position) { return static_cast<geometry_msgs::msg::Pose>(position); },
+      [&](const RelativeObjectPosition & position) { return static_cast<geometry_msgs::msg::Pose>(position); },
```
Comment (yamacir-kit) [inline openscenario/openscenario_interpreter/src/syntax/position.cpp:46]: Adjust the column alignment according to the style of the original code. Do not try to break the style to reduce diffs.

File: openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+#include <openscenario_interpreter/reader/element.hpp>
+#include <openscenario_interpreter/syntax/relative_object_position.hpp>
+
+namespace openscenario_interpreter
+{
+inline namespace syntax
+{
+RelativeObjectPosition::RelativeObjectPosition(const pugi::xml_node & node, Scope & scope)
+: orientation(readElement<Orientation>("Orientation", node, scope)),
+  reference(readAttribute<EntityRef>("entityRef", node, scope)),
```
Comment (yamacir-kit) [inline openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp:26]:
```suggestion
  entity_ref(readAttribute<String>("entityRef", node, scope)),
```

File: openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+RelativeObjectPosition::operator NativeLanePosition() const
+{
+  return convert<NativeLanePosition>(static_cast<NativeWorldPosition>(*this));
+}
+
+RelativeObjectPosition::operator NativeWorldPosition() const
+{
+  // get world origin pose
```
Comment (yamacir-kit) [inline openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp:52]: The commented code is not an operation that gets some value from somewhere else. Therefore, the word "get" is an inappropriate description.

```suggestion
  // Make world origin pose
```

File: openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  geometry_msgs::msg::Pose original_pose{};
+  original_pose.position.x = 0;
+  original_pose.position.y = 0;
+  original_pose.position.z = 0;
+  original_pose.orientation.x = 0;
+  original_pose.orientation.y = 0;
+  original_pose.orientation.z = 0;
+  original_pose.orientation.w = 1;
+
+  // get entity reference pose in world pose
```
Comment (yamacir-kit) [inline openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp:62]: The commented code is not an operation that gets some value from somewhere else. Therefore, the word "get" is an inappropriate description.

```suggestion
  // Make entity reference pose in world pose
```

File: openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+  const auto reference_pose = makeNativeRelativeWorldPosition(original_pose, reference);
+
+  // get offset point in reference pose
```
Comment (yamacir-kit) [inline openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp:65]: The commented code is not an operation that gets some value from somewhere else. Therefore, the word "get" is an inappropriate description.
In addition, comments such as this tautology are meaningless and should be deleted. The variable or function name should describe what is being processed, and the comment should be used to leave context that cannot be read from it. There is no need to repeat information that can be read at a glance through the code.

File: openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  geometry_msgs::msg::Point point_in_reference{};
+  point_in_reference.x = dx;
+  point_in_reference.y = dy;
+  point_in_reference.z = dz;
+
+  // calculate offset pose in world pose
+  const auto point_in_world = math::geometry::transformPoint(reference_pose, point_in_reference);
+  geometry_msgs::msg::Pose pose_in_world{};
+  pose_in_world.position = point_in_world;
+  pose_in_world.orientation = original_pose.orientation;
```
Comment (yamacir-kit) [inline openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp:75]:
```suggestion
  geometry_msgs::msg::Pose pose_in_world{};
  pose_in_world.position = math::geometry::transformPoint(reference_pose, point_in_reference);
  pose_in_world.orientation = original_pose.orientation;
```

_The use of overly general terms such as "calculate" in a description should be avoided. In a program, every line is a calculation._

**Review by yamacir-kit — CHANGES_REQUESTED** [review]

File: openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  original_pose.position.z = 0;
+  original_pose.orientation.x = 0;
+  original_pose.orientation.y = 0;
+  original_pose.orientation.z = 0;
+  original_pose.orientation.w = 1;
+
+  // Make entity reference pose in world pose
+  const auto reference_pose = makeNativeRelativeWorldPosition(original_pose, reference);
+
+  // get offset point in reference pose
```
Comment (kosuke55) [inline openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp:65]:
```suggestion
```

File: openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp (hunk 5 above)
Comment (kosuke55) [reply openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp]: I reffered [relative_world_position](https://github.com/tier4/scenario_simulator_v2/blob/master/openscenario/openscenario_interpreter/include/openscenario_interpreter/syntax/relative_world_position.hpp#L47) and if changing the name to `entity_ref` we need to change relative_worl_position and 
[applyTeleportAction](https://github.com/tier4/scenario_simulator_v2/blob/master/openscenario/openscenario_interpreter/src/syntax/teleport_action.cpp#L64) too. Is it better?

File: openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp (hunk 6 above)
Comment (kosuke55) [reply openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp]: > In addition, comments such as this tautology are meaningless and should be deleted. The variable or function name should describe what is being processed, and the comment should be used to leave context that cannot be read from it. There is no need to repeat information that can be read at a glance through the code.

Following the comment, removed tautology comments and reflect them in code
https://github.com/tier4/scenario_simulator_v2/pull/1030/commits/de9c67a8c425e8ec4dddf55460a40e37053cfd20

File: openscenario/openscenario_interpreter/src/syntax/position.cpp (hunk 3 above)
Comment (kosuke55) [reply openscenario/openscenario_interpreter/src/syntax/position.cpp:29]: fixed in https://github.com/tier4/scenario_simulator_v2/pull/1030/commits/7824f61848f356476a46d9e9555c9b820df1d519

File: openscenario/openscenario_interpreter/src/syntax/position.cpp (hunk 4 above)
Comment (kosuke55) [reply openscenario/openscenario_interpreter/src/syntax/position.cpp:46]: fixed https://github.com/tier4/scenario_simulator_v2/pull/1030/commits/7824f61848f356476a46d9e9555c9b820df1d519

File: openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp (hunk 5 above)
Comment (yamacir-kit) [reply openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp]: I understand why you used the data member name `reference`.

The `EntityRef` type data member name for `RelativeWorldPosition` being `reference` is odd given the current style of implementation of syntax-related types. I checked and the code in question was [very old code implemented by me on 12/04/2020](https://github.com/tier4/scenario_simulator_v2/commit/30aa81d7c958bb02cc2461e215a1a1071a354e88). Perhaps at that time I had not defined a (private) rule that the structure of syntax-related classes should match well with the OpenSCENARIO standard.

Therefore, the answer to your question is "yes, that is better".

Comment (kosuke55) [reply openscenario/openscenario_interpreter/src/syntax/relative_object_position.cpp]: ok, thanks! Fixed in https://github.com/tier4/scenario_simulator_v2/pull/1030/commits/0dfdb61c8df0b18a1e384b50f58a1d22562bfe5b

File: .github/workflows/custom_spell.json [hunk 11]
```diff
@@ -3,7 +3,8 @@
     "*.drawio",
     "*/**/CHANGELOG.rst",
     "external/lanelet2_matching/**",
-    "openscenario/openscenario_utility/openscenario_utility/resources/*.xsd"
+    "openscenario/openscenario_utility/openscenario_utility/resources/*.xsd",
+    "test_runner/scenario_test_runner/scenario/spawn_relative_object_position.yaml"
```
Comment (kosuke55) [inline .github/workflows/custom_spell.json:7]: @yamacir-kit 
I wrote in the json file that "euclidian" is misspelled, but it seems that comments on json are not allowed. The comment itself is also treated as a misspelling. Is it OK if I don't comment?

image

Comment (kosuke55) [reply .github/workflows/custom_spell.json:7]: @HansRobo 
I have received a forward to you could you please your advice?

Comment (HansRobo) [reply .github/workflows/custom_spell.json]: close with https://github.com/tier4/scenario_simulator_v2/pull/1030/commits/8e6bca7532b8798a3e4fc8f2f0e34abbb3ee3f91

Comment (kosuke55) [reply .github/workflows/custom_spell.json]: thanks and add reason for disabling
https://github.com/tier4/scenario_simulator_v2/pull/1030/commits/9578caacd756c45529099cea657e8c1a02765861

**Comment by kosuke55** [comment]
@yamacir-kit 
I finished the degradation check and it was ok! 
https://github.com/tier4/sim_evaluation_tools/issues/159

**Review by yamacir-kit — CHANGES_REQUESTED** [review]

**Comment by kosuke55** [comment]
@yamacir-kit 
thnaks for your suggestion, I updated `dy` `0` -> `-1` and checked it passed. [758089e](https://github.com/tier4/scenario_simulator_v2/pull/1030/commits/758089e4c4cddcc99ec89564aefc0292070e69c6)

https://github.com/tier4/scenario_simulator_v2/assets/39142679/58f626b9-26da-49f4-8c46-40f0ccb6116d

File: docs/developer_guide/OpenSCENARIOSupport.md [hunk 12]
```diff
@@ -376,7 +376,7 @@ See also section [Scoping](#scoping).
 | RelativeWorldPosition  | Unsupported |
 | RelativeObjectPosition | Unsupported |
 | RoadPosition           | Unsupported |
-| RelativeRoadPosition   | Unsupported |
+| RelativeRoadPosition   |      ✔      |
```
Comment (yamacir-kit) [inline docs/developer_guide/OpenSCENARIOSupport.md:379]: What this pull request introduces is RelativeObjectPosition, not RelativeRoadPosition.

**Review by yamacir-kit — CHANGES_REQUESTED** [review]

File: docs/developer_guide/OpenSCENARIOSupport.md (hunk 12 above)
Comment (kosuke55) [reply docs/developer_guide/OpenSCENARIOSupport.md]: oh, sorry for the mistake, fixed in https://github.com/tier4/scenario_simulator_v2/pull/1030/commits/7ef4ad3287dd981d61006b2d6f5ff0a90d0d0f64

**Comment by yamacir-kit** [comment]
@kosuke55 
It should almost certainly be OK, but please run the regression test again to check the impact of [7fc19e1](https://github.com/tier4/scenario_simulator_v2/pull/1030/commits/7fc19e1a1949146e0c57787391c2db12ff5aedf6).

**Comment by kosuke55** [comment]
@yamacir-kit 
I checked the degradation test again and it was ok.
https://github.com/tier4/sim_evaluation_tools/issues/171

**Review by yamacir-kit — APPROVED** [review]

**Pull request merged**